pkg/controller/controller_instance.go

```diff
@@ -329,6 +353,86 @@ func (c *controller) initOrphanMitigationCondition(instance *v1beta1.ServiceInst
 	return false, nil
 }
 
+// setNextProvisionRetryTime adds the instance name to the map recording the next earliest time to
+// retry provisioning
+func (c *controller) setNextProvisionRetryTime(instance *v1beta1.ServiceInstance) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(instance)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", instance, err)
+		return
+	}
+	duration := c.provisionRetryQueue.rateLimiter.When(key)
+	c.provisionRetryQueue.mutex.Lock()
+	c.provisionRetryQueue.retryTime[key] = time.Now().Add(duration)
+	glog.V(4).Infof("provisionRetry for %s after %v", key, duration)
+	c.provisionRetryQueue.mutex.Unlock()
+}
+
+//  delayProvisionRetry returns a duration which should be observed before attempting to provision
+func (c *controller) getDelayForProvisionRetry(instance *v1beta1.ServiceInstance) time.Duration {
+	c.purgeExpiredRetryEntries()
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(instance)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", instance, err)
+		return 0
+	}
+	c.provisionRetryQueue.mutex.RLock()
+	defer c.provisionRetryQueue.mutex.RUnlock()
+	now := time.Now()
+	if t := c.provisionRetryQueue.retryTime[key]; t.After(now) {
+		return t.Sub(now)
+	}
+	return 0
+}
+
+// purgeExpiredRetryEntries periodically clears entries from the map that have an expired
+// retry time.
+func (c *controller) purgeExpiredRetryEntries() {
+	now := time.Now()
+
+	// run periodically ensuring we don't prematurely purge any entries
+	timeToPurge := c.provisionRetryQueue.queueLastCleaned.Add(maxBrokerProvisioningRetryDelay * 2)
+	if now.Before(timeToPurge) {
+		return
+	}
+	c.provisionRetryQueue.mutex.Lock()
+	defer c.provisionRetryQueue.mutex.Unlock()
+
+	// Ensure we only purge items that aren't being acted on by retries, this
+	// shouldn't have much work to do but we want to be certain this queue
+	// doesn't get overly large. Entries are removed one by one when deleted
+	// (not orphan mitigation) or successfully provisioned, this function ensures
+	// all others get purged eventually.  Due to queues and potential delays,
+	// only remove entries that are at least maxBrokerProvisioningRetryDelay
+	// past next retry time to ensure entries are not prematurely removed
+	overDue := now.Add(-maxBrokerProvisioningRetryDelay)
+	purgedEntries := 0
+	for k := range c.provisionRetryQueue.retryTime {
+		if c.provisionRetryQueue.retryTime[k].Before(overDue) {
+			glog.V(5).Infof("removed %s from provisionRetry map which had retry time of %v", k, c.provisionRetryQueue.retryTime[k])
```

Comment:
I prefer to cache the value in the var if it's being used twice (`c.provisionRetryQueue.retryTime[k]`).